File: frontend/src/api/business/materialCount.js
import request from '../../utils/request';

/**
 * 材料盘点API
 */

// 获取材料盘点列表
export const getMaterialCounts = (params) => {
  return request({
    url: '/tenant/business/inventory/material-count/material-count-orders',
    method: 'get',
    params
  });
};

// 创建材料盘点
export const createMaterialCount = (data) => {
  return request({
    url: '/tenant/business/inventory/material-count/material-count-orders',
    method: 'post',
    data
  });
};

// 获取材料盘点详情
export const getMaterialCountById = (id) => {
  return request({
    url: `/tenant/business/inventory/material-count/material-count-orders/${id}`,
    method: 'get'
  });
};

// 更新材料盘点
export const updateMaterialCount = (id, data) => {
  return request({
    url: `/tenant/business/inventory/material-count/material-count-orders/${id}`,
    method: 'put',
    data
  });
};

// 删除材料盘点
export const deleteMaterialCount = (id) => {
  return request({
    url: `/tenant/business/inventory/material-count/material-count-orders/${id}`,
    method: 'delete'
  });
};

// 确认材料盘点
export const confirmMaterialCount = (id) => {
  return request({
    url: `/tenant/business/inventory/material-count/material-count-orders/${id}/approve`,
    method: 'post'
  });
};

// 获取材料库存
export const getMaterialInventory = (params) => {
  return request({
    url: '/tenant/business/inventory/inventories',
    method: 'get',
    params
  });
};

// 导出材料盘点
export const exportMaterialCount = (params) => {
  return request({
    url: '/tenant/business/inventory/material-count/material-count-orders/export',
    method: 'get',
    params,
    responseType: 'blob'
  });
};

// 获取仓库材料库存
export const getWarehouseMaterialInventory = (warehouseId) => {
  return request.get(`/tenant/business/inventory/inventories?warehouse_id=${warehouseId}`);
};

// 获取仓库列表（用于选择仓库，只获取材料仓）
export const getWarehouses = (params = {}) => {
  return request.get('/tenant/base-archive/base-data/warehouses/options', { 
    params: { 
      ...params,
      warehouse_type: 'material' // 只获取材料仓
    } 
  });
};

// 获取盘点计划列表（实际就是盘点列表）
export const getMaterialCountPlans = (params = {}) => {
  return request.get('/tenant/business/inventory/material-count/material-count-orders', { params });
};

// 创建盘点计划（实际就是创建盘点）
export const createMaterialCountPlan = (data) => {
  return request.post('/tenant/business/inventory/material-count/material-count-orders', data);
};

// 获取盘点计划详情（实际就是盘点详情）
export const getMaterialCountPlan = (planId) => {
  return request.get(`/tenant/business/inventory/material-count/material-count-orders/${planId}`);
};

// 获取盘点记录
export const getMaterialCountRecords = (planId, params = {}) => {
  return request.get(`/tenant/business/inventory/material-count/material-count-orders/${planId}/records`, { params });
};

// 更新盘点记录
export const updateMaterialCountRecord = (planId, recordId, data) => {
  return request.put(`/tenant/business/inventory/material-count/material-count-orders/${planId}/records/${recordId}`, data);
};

// 开始盘点
export const startMaterialCountPlan = (planId) => {
  return request.post(`/tenant/business/inventory/material-count/material-count-orders/${planId}/start`);
};

// 完成盘点
export const completeMaterialCountPlan = (planId) => {
  return request.post(`/tenant/business/inventory/material-count/material-count-orders/${planId}/complete`);
};

// 调整库存
export const adjustMaterialCountInventory = (planId, data = {}) => {
  return request.post(`/tenant/business/inventory/material-count/material-count-orders/plans/${planId}/adjust`, data);
};

// 统一导出API对象
export const materialCountApi = {
  // 基础盘点API
  getMaterialCounts,
  createMaterialCount,
  updateMaterialCount,
  deleteMaterialCount,
  getMaterialCountById,
  confirmMaterialCount,
  getMaterialInventory,
  exportMaterialCount,
  // 盘点计划API
  getMaterialCountPlans,
  createMaterialCountPlan,
  getMaterialCountPlan,
  getMaterialCountRecords,
  updateMaterialCountRecord,
  startMaterialCountPlan,
  completeMaterialCountPlan,
  adjustMaterialCountInventory,
  getWarehouseMaterialInventory,
  getWarehouses
};

// 保持向后兼容的默认导出
export default materialCountApi; 

File: backend/app/api/tenant/business/inventory/material_count.py
# type: ignore
# pyright: reportGeneralTypeIssues=false
# pyright: reportAttributeAccessIssue=false
# pyright: reportOptionalMemberAccess=false
"""
材料盘点管理 API

提供材料盘点的完整管理功能：
- 材料盘点列表查询和创建
- 材料盘点详情获取和更新
- 材料盘点删除和状态管理
- 材料盘点审核、执行、取消流程
"""

from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from app.api.tenant.routes import tenant_required
import logging

# 设置蓝图
bp = Blueprint('material_count', __name__)
logger = logging.getLogger(__name__)

# ==================== 材料盘点管理 ====================

@bp.route('/material-count-orders', methods=['GET'])
@jwt_required()
@tenant_required
def get_material_count_orders():
    """获取材料盘点列表"""
    try:
        # 获取查询参数
        page = int(request.args.get('page', 1))
        page_size = min(int(request.args.get('page_size', 20)), 100)
        count_number = request.args.get('count_number')
        warehouse_id = request.args.get('warehouse_id')
        status = request.args.get('status')
        approval_status = request.args.get('approval_status')
        start_date = request.args.get('start_date')
        end_date = request.args.get('end_date')
        
        
        # 获取材料盘点列表
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        result = service.get_material_count_list(
            warehouse_id=warehouse_id,
            status=status,
            start_date=start_date,
            end_date=end_date,
            search=count_number,
            page=page,
            page_size=page_size
        )
        
        return jsonify({
            'success': True,
            'data': result
        })
        
    except Exception as e:
        logger.error(f"获取材料盘点列表失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<count_id>', methods=['GET'])
@jwt_required()
@tenant_required
def get_material_count_order(count_id):
    """获取材料盘点详情"""
    try:
        claims = get_jwt()
        tenant_id = claims.get('tenant_id')
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        count = service.get_material_count_by_id(count_id)
        
        if not count:
            return jsonify({'error': '材料盘点不存在'}), 404
        
        return jsonify({
            'success': True,
            'data': count
        })
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 404
    except Exception as e:
        logger.error(f"获取材料盘点详情失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders', methods=['POST'])
@jwt_required()
@tenant_required
def create_material_count_order():
    """创建材料盘点"""
    try:
        current_user_id = get_jwt_identity()

        data = request.get_json()
        
        if not data:
            return jsonify({'error': '请求数据不能为空'}), 400
        
        # 验证必填字段
        if not data.get('warehouse_id'):
            return jsonify({'error': '仓库ID不能为空'}), 400
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        count = service.create_material_count(data, current_user_id)
        
        return jsonify({
            'success': True,
            'data': count,
            'message': '材料盘点创建成功'
        }), 201
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        logger.error(f"创建材料盘点失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<count_id>/start', methods=['POST'])
@jwt_required()
@tenant_required
def start_material_count_order(count_id):
    """开始材料盘点"""
    try:
        current_user_id = get_jwt_identity()
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        result = service.start_material_count(count_id, current_user_id)
        
        return jsonify({
            'success': True,
            'data': result,
            'message': '盘点已开始'
        })
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        logger.error(f"开始材料盘点失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<count_id>/complete', methods=['POST'])
@jwt_required()
@tenant_required
def complete_material_count_order(count_id):
    """完成材料盘点"""
    try:
        current_user_id = get_jwt_identity()
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        result = service.complete_material_count(count_id, current_user_id)
        
        return jsonify({
            'success': True,
            'data': result,
            'message': '盘点已完成'
        })
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        logger.error(f"完成材料盘点失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<order_id>/adjust', methods=['POST'])
@jwt_required()
@tenant_required
def adjust_material_count_inventory(order_id):
    """调整材料盘点库存"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json() or {}
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        result = service.adjust_material_count_inventory(order_id, current_user_id)
        
        return jsonify({
            'success': True,
            'data': result,
            'message': '库存调整成功'
        })
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        logger.error(f"调整材料盘点库存失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<count_id>/records', methods=['GET'])
@jwt_required()
@tenant_required
def get_material_count_records(count_id):
    """获取材料盘点记录"""
    try:
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        records = service.get_material_count_records(count_id)
        
        return jsonify({
            'success': True,
            'data': records
        })
        
    except Exception as e:
        logger.error(f"获取材料盘点记录失败: {str(e)}")
        return jsonify({'error': str(e)}), 500


@bp.route('/material-count-orders/<count_id>/records/<record_id>', methods=['PUT'])
@jwt_required()
@tenant_required
def update_material_count_record(count_id, record_id):
    """更新材料盘点记录"""
    try:
        current_user_id = get_jwt_identity()
        data = request.get_json()
        
        if not data:
            return jsonify({'error': '请求数据不能为空'}), 400
        
        from app.services.business.inventory.material_count_service import MaterialCountService
        service = MaterialCountService()
        record = service.update_material_count_record(record_id, current_user_id, data)
        
        return jsonify({
            'success': True,
            'data': record,
            'message': '盘点记录更新成功'
        })
        
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        logger.error(f"更新材料盘点记录失败: {str(e)}")
        return jsonify({'error': str(e)}), 500
